posthog/temporal/batch_exports/spmc.py

```diff
@@ -305,9 +308,8 @@ async def start(
 
                 records_count += writer.records_since_last_flush
 
-                if multiple_files:
-                    await writer.close_temporary_file()
-                    writer._batch_export_file = await asyncio.to_thread(writer.create_temporary_file)
+                if multiple_files or (max_file_size_bytes > 0 and writer.bytes_total >= max_file_size_bytes):
+                    await writer.hard_flush()
```

Comment:
I'm afraid we may be late in flushing here if a record batch contains records exceeding `max_file_size_bytes` by double the amount.

E.g.:
* Read record batch containing records with 100mb.
* `max_file_size_bytes` is set to 50mb.
* Record batch is written a few lines above to a local file now weighing 100mb.
* We reach here, naturally 100mb > 50mb, so we flush.
* The user receives a 100mb file, instead of two 50mb, which they likely expected.

Reply:
The crux of the issue is that we can't know the size of a record batch before reading it from clickhouse. You can know it before writing it to local filesystem though, so maybe we should check there and split the record batch before writing it.

Reply:
Yes you're right. I was working under the assumption that it would be 'good enough' (the original aim here is to not generate very large files, eg ones that are several GB, so assumed it doesn't matter too much if a file is 600MB rather than 500MB). But maybe we ought to enforce it a bit better?

> You can know it before writing it to local filesystem though, so maybe we should check there and split the record batch before writing it.

I could do something like this. I will see how easy this is to do.

Reply:
> Yes you're right. I was working under the assumption that it would be 'good enough' (the original aim here is to not generate very large files, eg ones that are several GB, so assumed it doesn't matter too much if a file is 600MB rather than 500MB). But maybe we ought to enforce it a bit better?

Yeah, this is likely to be more of an issue at lower max file size bytes. Anything more than a couple hundred MB and its very likely clickhouse won't give us that much data in one record batch.